File: tools/browser-agent/browser_agent/humanize.py
"""人类节奏：动作延迟、打字间隔、限速。纯函数 / 可注入时钟，便于单测。"""
from __future__ import annotations

import math
import random
import time
from collections import deque
from typing import Callable, Deque, Dict


def action_delay(lo: float = 0.4, hi: float = 1.5) -> float:
    """两次动作间的随机延迟（秒），偏向区间中段的对数正态抖动。"""
    mu = (math.log(lo) + math.log(hi)) / 2
    sigma = (math.log(hi) - math.log(lo)) / 4
    val = math.exp(random.gauss(mu, sigma))
    return max(lo, min(hi, val))


def typing_intervals(text: str, base: float = 0.08) -> list[float]:
    """逐字输入的间隔列表，模拟人类打字节奏（含偶发停顿）。"""
    out: list[float] = []
    for ch in text:
        jitter = random.uniform(0.6, 1.6)
        pause = 0.25 if ch == " " and random.random() < 0.2 else 0.0
        out.append(min(0.5, max(0.02, base * jitter + pause)))
    return out


class RateLimiter:
    """滑动窗口限速：每个 key（站点）独立预算。"""

    def __init__(self, max_per_window: int, window_s: float,
                 _now: Callable[[], float] | None = None) -> None:
        self._max = max_per_window
        self._window = window_s
        self._now = _now or time.monotonic
        # 按 key（站点）存命中时间戳；本场景 key 是少量固定站点，集合有界，不做过期清理。
        self._hits: Dict[str, Deque[float]] = {}

    def allow(self, key: str) -> bool:
        now = self._now()
        dq = self._hits.setdefault(key, deque())
        while dq and now - dq[0] > self._window:
            dq.popleft()
        if len(dq) >= self._max:
            return False
        dq.append(now)
        return True

File: tools/browser/src/humanize.js
/** 人类节奏：动作延迟、打字间隔、限速、鼠标轨迹。纯函数 / 可注入时钟。 */

function gauss(mu, sigma) {
  // Box-Muller
  const u = 1 - Math.random();
  const v = Math.random();
  return mu + sigma * Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * v);
}

/** 两次动作间随机延迟（秒），偏区间中段的对数正态。 */
export function actionDelay(lo = 0.4, hi = 1.5) {
  const mu = (Math.log(lo) + Math.log(hi)) / 2;
  const sigma = (Math.log(hi) - Math.log(lo)) / 4;
  return Math.max(lo, Math.min(hi, Math.exp(gauss(mu, sigma))));
}

/** 逐字输入间隔（秒），含偶发空格停顿。 */
export function typingIntervals(text, base = 0.08) {
  const out = [];
  for (const ch of text) {
    const jitter = 0.6 + Math.random();
    const pause = ch === " " && Math.random() < 0.2 ? 0.25 : 0;
    out.push(Math.min(0.5, Math.max(0.02, base * jitter + pause)));
  }
  return out;
}

/** 鼠标从 from 到 to 的分段轨迹（含微抖动），共 steps+1 个点，端点精确。 */
export function mousePath(from, to, steps = 12) {
  const pts = [{ x: from.x, y: from.y }];
  for (let i = 1; i < steps; i++) {
    const t = i / steps;
    const jx = (Math.random() - 0.5) * 4;
    const jy = (Math.random() - 0.5) * 4;
    pts.push({
      x: from.x + (to.x - from.x) * t + jx,
      y: from.y + (to.y - from.y) * t + jy,
    });
  }
  pts.push({ x: to.x, y: to.y });
  return pts;
}

/** 滑动窗口限速，每 key 独立。可注入时钟（毫秒/秒一致即可）。 */
export class RateLimiter {
  constructor(maxPerWindow, windowS, now = () => Date.now() / 1000) {
    this._max = maxPerWindow;
    this._window = windowS;
    this._now = now;
    this._hits = new Map();
  }
  allow(key) {
    const t = this._now();
    const dq = this._hits.get(key) ?? [];
    while (dq.length && t - dq[0] > this._window) dq.shift();
    if (dq.length >= this._max) {
      this._hits.set(key, dq);
      return false;
    }
    dq.push(t);
    this._hits.set(key, dq);
    return true;
  }
}

/** 睡眠 helper（秒）。 */
export const sleep = (s) => new Promise((r) => setTimeout(r, s * 1000));
